crates/ruff_linter/src/rules/pyupgrade/rules/yield_in_for_loop.rs

```diff
@@ -100,13 +100,29 @@ pub(crate) fn yield_in_for_loop(checker: &mut Checker, stmt_for: &ast::StmtFor)
         checker
             .semantic()
             .current_scope()
-            .get_all(name.id.as_str())
-            .any(|binding_id| {
-                let binding = checker.semantic().binding(binding_id);
+            .get(name.id.as_str())
+            .map(|mut binding_id| {
+                let mut binding = checker.semantic().binding(binding_id);
+
+                // If the binding is an unbound kind shadowing another binding, resolve it to the
+                // shadowed binding.
+                while binding.is_unbound() {
+                    let Some(shadowed_id) = checker
+                        .semantic()
+                        .current_scope()
+                        .shadowed_binding(binding_id)
+                    else {
+                        break;
+                    };
+                    binding_id = shadowed_id;
+                    binding = checker.semantic().binding(binding_id);
+                }
+
                 binding.references.iter().any(|reference_id| {
                     checker.semantic().reference(*reference_id).range() != name.range()
                 })
             })
+            .unwrap_or(false)
```

Comment:
Nit: You could avoid rolling your own shadow-binding traversal by keeping `get_all` and use `find`

```rust
        checker
            .semantic()
            .current_scope()
            .get_all(&name.id)
            // If the binding is an unbound kind shadowing another binding, resolve it to the
            // shadowed binding.
            .find(|&id| !checker.semantic().binding(id).is_unbound())
            .map(|binding_id| {
                let binding = checker.semantic().binding(binding_id);
                binding.references.iter().any(|reference_id| {
                    checker.semantic().reference(*reference_id).range() != name.range()
                })
            })
            .unwrap_or_default()
```

Reply:
Thanks! Not sure why I thought that was different.